File: awa/sdk_dist/csharp/Awa.Client/Utilities/Workflow/ExecuteBamlTransformBatchWorkflow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Awa.Client.Constants;
using Awa.Client.Models;
using Awa.Client.Utilities.Activity;

namespace Awa.Client.Utilities.Workflow
{
    public static partial class WorkflowUtilities
    {
        /// <summary>
        /// Execute multiple BAML transformations in batch using a specified function.
        /// </summary>
        /// <param name="bamlRequestsByKey">Dictionary mapping keys to BAML transform params. Each key will have its corresponding request processed.</param>
        /// <param name="bamlPath">Optional path to the BAML file (will be read and added as content to the transform params).</param>
        /// <returns>Dictionary mapping the same keys to their transformation results.</returns>
        public static async Task<Dictionary<string, object?>> ExecuteBamlTransformBatchWorkflow(
            Dictionary<string, TransformParams> bamlRequestsByKey,
            string? bamlPath = null)
        {
            if (bamlRequestsByKey == null || bamlRequestsByKey.Count == 0)
            {
                return new Dictionary<string, object?>();
            }

            if (!string.IsNullOrEmpty(bamlPath))
            {
                var bamlContent = await ActivityUtilities.ReadFileActivity(bamlPath);
                foreach (var transformParams in bamlRequestsByKey.Values)
                {
                    transformParams.BamlContent = bamlContent;
                }
            }

            var bamlFunctionName = bamlRequestsByKey.Values.First().BamlFunctionName;
            return await Temporalio.Workflows.Workflow.ExecuteChildWorkflowAsync<Dictionary<string, object?>>(
                AwaConstants.WorkflowTransformBatch,
                new object?[] { new { params_by_key = bamlRequestsByKey, timeout_seconds = AwaConstants.BamlTimeoutSeconds } },
                new()
                {
                    TaskQueue = AwaConstants.AwaDefaultTaskQueue,
                    Id = $"{bamlFunctionName}-{Temporalio.Workflows.Workflow.Info.WorkflowId}"
                });
        }
    }
}

File: awa/sdk_dist/python/awa/client/utils/workflow/execute_baml_transform_batch_workflow.py
from pathlib import Path
from typing import Any

from temporalio import workflow

from awa.client import constants
from awa.client.models.transform_params import TransformParams
from awa.client.utils.activity.read_file_activity import read_file_activity


async def execute_baml_transform_batch_workflow(
    baml_requests_by_key: dict[str, TransformParams],
    baml_path: str | Path | None = None,
) -> dict[str, Any]:
    """Execute multiple BAML transformations in batch using a specified function.

    Args:
        baml_requests_by_key: Dictionary mapping keys to BAML transform params.
            Each key will have its corresponding request processed.
        baml_path: Optional path to the BAML file (will be read and added as content to the tranform params).

    Returns:
        dict[str, Any]: Dictionary mapping the same keys to their transformation results.

    """
    if not baml_requests_by_key or len(baml_requests_by_key) == 0:
        return {}

    if baml_path:
        baml_content = await read_file_activity(str(baml_path))
        for transform_params in baml_requests_by_key.values():
            transform_params.baml_content = baml_content

    baml_function_name = next(iter(baml_requests_by_key.values())).baml_function_name
    return await workflow.execute_child_workflow(
        workflow=constants.WORKFLOW_TRANSFORM_BATCH,
        arg={"params_by_key": baml_requests_by_key, "timeout_seconds": constants.BAML_TIMEOUT_SECONDS},
        task_queue=constants.AWA_DEFAULT_TASK_QUEUE,
        id=f"TransformBatch-{workflow.uuid4()}",
        static_summary=f"TransformBatch-{baml_function_name} ({len(baml_requests_by_key)})",
        static_details="\n".join(baml_requests_by_key.keys()),
    )
